CIP-array-builtin/README.md

```diff
@@ -0,0 +1,286 @@
+---
+
+## Abstract
+
+We propose an array builtin type for Plutus Core. This type will have constant-time lookup,
+which is a useful feature that is otherwise not possible to achieve. 
+
+## Motivation: why is this CIP necessary?
+
+The first part of [CPS-0013][1] outlines in great detail the motivation for introducing this
+new builtin type.
+
+To summarize, it is currently not possible to write a data structure with constant-time
+lookup in Plutus Core. We propose to solve this problem by introducing an array type into
+Plutus Core's builtin language, as it is the standard example of a data structure
+with this property, and it is a key component of many classical algorithms and data
+structures.
+
+Access to an array type would provide significant performance improvement opportunities to
+users of Plutus Core, since currently they must rely on suboptimal data structures such as
+lists for looking up elements inside a collection.
+
+## Specification
+
+We add the following builtin type: `Array` of kind `Type -> Type` representing
+a one-dimensional array type with indices of type `Integer`. Elements are indexed
+consecutively starting from `0`.
+
+***Note***: Here `Type` is the universe of all _builtin_ types, since we do not consider
+types formed out of applying builtin types to arbitrary types to be inhabitted.
+
+The `Array` builtin type should be implemented with a fixed-size immutable array structure
+that has constant-time lookup.
+
+We add the following builtin functions:
+
+1. `indexArray` of type `forall a . Integer -> Array a -> a`.
+    - It returns the element at the given index in the array, or fails with an error if the
+    index is outside the bounds of the array.
+    - It uses constant time and constant memory.
+2. `lengthOfArray` of type `forall a . Array a -> Integer`
+    - It returns the length of the array.
+    - It uses constant time and constant memory.
+3. `listToArray` of type `forall a . List a -> Array a`
+    - It converts the argument builtin list into a builtin array.
+    - It uses linear time and linear memory.
+
+### Binary serialisation and deserialisation
+
+As with all Plutus Core builtin types, arrays must have a fixed binary representation.
+
+For arrays we have chosen the following representation, based on the one currently
+implemented in the [flat][8] encoding for the [Haskell `Array` type][9].
+
+We define the result of encoding a Plutus Core array of type `forall a . Array a`
+and with length `n`, as follows:
+- Let:
+    - `Ea` be an encoding function for `a`, i.e. it is of type `a -> ByteString`.
+    - `Ei` be an encoding function for `Integer`
+    - `++` be bytestring concatenation
+    - `bin(i)` be the binary representation for any number `i`.
+- The first bits in the resulting sequence will be `Ei(0) ++ Ei(n-1)`, that is the
+encoding of the first index followed by the encoding of the last.
+- As binary, arrays are split into _blocks_ of `255` elements. For each block, the first
```

Comment:
> For each block, the first bits in the block encoding can be decided as follows

I don't quite understand what that means.

Reply:
Sorry, I'll try to reformulate. The idea is that the encoding of each block starts with the number of elements that block contains.